src/atc/eh/EventHubCapture.py

```diff
@@ -0,0 +1,210 @@
+
+        # discover each part
+        for c in self.partitioning:
+            # loop over items in path
+            value = None
+            for file_info in dbutils.fs.ls(self.path + "/" + partition_path):
+                # validate
+                if not file_info.path.startswith("dbfs:"):
+                    raise AtcEhLogicException("Unexpected file paths")
+
+                if not (
+                    file_info.name.startswith(c + "=") and file_info.name.endswith("/")
+                ):
+                    continue
+                extraction = file_info.name[2:-1]
+
+                # keep only the lowest
+                if value is None or int(extraction) < int(value):
+                    value = extraction
+
+            # done looping over items.
+            if value is None:
+                # There is probably no data, yet.
+                raise AtcEhInitException("unable to discover first partition")
+
+            if partition_path:
+                partition_path += "/"
+            partition_path += f"{c}={value}"
+            partition[c] = value
+
+        # we have discovered all parts of the partition.
+        Spark.get().sql(
+            f"""ALTER TABLE {self.name} ADD PARTITION ({
+            ", ".join(f"{c}='{v}'" for c, v in partition.items())
+            })"""
+        )
+
+        # return the partition in path form so that the repair can continue
+        return partition_path
+
+    def _get_max_partition(self):
+        """get the current highest partition either
+        1. from memory
+        2. read the partition table
+        3. initialize from disk
+        """
+        if self.max_partition is None:
+            # memory access failed.
+            # try reading partition table
+            try:
+                max_p = (
+                    Spark.get()
+                    .sql(f"SHOW PARTITIONS {self.name}")
+                    .groupBy()
+                    .agg(f.max("partition"))
+                    .collect()
+                )[0][0]
+                # value will be None if there is no value
+            except AnalysisException:
+                # no such table. Better create one
+                self._create_table()
+                # new table has no partitions
+                max_p = None
+
+            if not max_p:
+                # reading partition metadata failed.
+                # initialize first partition from disk.
+                max_p = self._discover_first_partition()
+
+            # we got the partition in the form y=2022/m=04/d=04
+            # extract the parts
+            pat = re.compile("/".join(f"{c}=(?P<{c}>\\d+)" for c in self.partitioning))
+            match = pat.match(max_p)
+            self.max_partition = datetime(
+                year=int(match.group("y")),
+                month=int(match.group("m")),
+                day=int(match.group("d")),
+                hour=int(match.group("h")) if "h" in self.partitioning else 0,
+            )
```

Comment:
when using "datetime" should we be aware of the UTC-problems? (difference between cluster vs. local python time)

Reply:
I have thought about it, but nowhere in this file does it matter what timezone the object is in, so I decided to go with a 'naive' object.